drivers/adc/adc_ad405x.c

```diff
@@ -174,9 +179,94 @@ struct adc_ad405x_data {
 	struct gpio_callback gpio0_cb;
 	struct k_sem sem_drdy;
 	uint8_t has_drdy;
+#endif /* CONFIG_AD405X_TRIGGER */
+#ifdef CONFIG_AD405X_STREAM
+	struct rtio_iodev_sqe *sqe;
+	struct rtio *rtio_ctx;
+	struct rtio_iodev *iodev;
+	uint64_t timestamp;
+	struct rtio *r_cb;
+	uint32_t adc_sample;
+	uint8_t data_ready_gpio;
+#if DT_HAS_CHOSEN(zephyr_adc_clock)
+	const struct device *timer_dev;
+#else
+	struct k_timer sample_timer;
 #endif
+#endif /* CONFIG_AD405X_STREAM */
+};
+
+#ifdef CONFIG_AD405X_STREAM
+#include <zephyr/drivers/counter.h>
+
+#define AD405X_DEF_SAMPLING_PERIOD 10000U /* 10ms */
+
+/** AD405X qscale modes */
+enum ad405x_qscale_modes {
+	AD4050_6_12B_MODE = 0,
+	AD4050_6_14B_MODE = 1,
+	AD4052_8_16B_MODE = 2,
+	AD4052_8_20B_MODE = 3,
 };
 
+struct adc_ad405x_fifo_data {
+	uint8_t is_fifo: 1;
+	uint8_t ad405x_qscale_mode: 2;
+	uint8_t diff_mode: 1;
+	uint8_t empty: 1;
+	uint8_t res: 3;
+	uint16_t vref_mv;
+	uint64_t timestamp;
+} __attribute__((__packed__));
+
+static void ad405x_stream_irq_handler(const struct device *dev);
+static int ad405x_conv_start(const struct device *dev);
+
+#if DT_HAS_CHOSEN(zephyr_adc_clock)
+static void timer_alarm_handler(const struct device *counter_dev, uint8_t chan_id,
+	uint32_t ticks, void *user_data)
+{
+	struct adc_ad405x_data *data = (struct adc_ad405x_data *)user_data;
+
+	ad405x_conv_start(data->dev);
+}
+#else
+static void sample_timer_handler(struct k_timer *timer)
+{
+	struct adc_ad405x_data *data = CONTAINER_OF(timer, struct adc_ad405x_data,
+						sample_timer);
+
+	ad405x_conv_start(data->dev);
+}
+#endif
+static void ad405x_timer_init(const struct device *dev)
+{
+	struct adc_ad405x_data *data = dev->data;
+
+#if DT_HAS_CHOSEN(zephyr_adc_clock)
+	counter_start(data->timer_dev);
+#else
+	k_timer_init(&data->sample_timer, sample_timer_handler, NULL);
+#endif
+}
+
+static void ad405x_timer_start(const struct device *dev)
+{
+	struct adc_ad405x_data *data = dev->data;
+	const struct adc_ad405x_config *cfg_405 = (const struct adc_ad405x_config *)dev->config;
+#if DT_HAS_CHOSEN(zephyr_adc_clock)
+	struct counter_alarm_cfg alarm_cfg = {.flags = 0,
+		.ticks = counter_us_to_ticks(data->timer_dev, (uint64_t)cfg_405->sampling_period),
+		.callback = timer_alarm_handler,
+		.user_data = data};
+
+	counter_set_channel_alarm(data->timer_dev, 0, &alarm_cfg);
+#else
+	k_timer_start(&data->sample_timer, K_USEC(cfg_405->sampling_period), K_NO_WAIT);
```

Comment:
IMO - This API is missing a way to configure the streaming mode. For instance: setting the watermark level, or the DRDY period (like this).

A default DTS property shouldn't be the only way we'd set these things.

Reply:
This is definitely something that could/should be looked at as configuration, the adc configuration today mostly comes int he adc_sequence struct (interval_us, channels enabled, resolution, oversampling, etc)

Arguably adc_sequence could be changed slightly to extract all the configuration away from the buffer pointer and len pair that are in it to make this a bit cleaner.

That would likely require updating many drivers. Please file an issue to make this a requirement for stabilizing this ADC stream API and I'd be ok with the existing setup.